Adds one-shot mDNS resolution

benbrittain opened the pull request:
RFC 6762 Section 5.1 specifies a one-shot multicast DNS query. This
query has minimal differences from a standard DNS query, mostly just
using a multicast address and a different port (5353 vs 53).

A fully standards compliant mDNS implementation would use UDP source
port 5353 as well to issue queries, however we MUST NOT use that port
and continue using an ephemeral port until features such as service
discovery are implemented.

This change also allows specifying what kind of DNS query we wish to
perform.

https://www.rfc-editor.org/rfc/rfc6762#section-5.1

benbrittain committed:
Adds one-shot mDNS resolution

RFC 6762 Section 5.1 specifies a one-shot multicast DNS query. This
query has minimal differences from a standard DNS query, mostly just
using a multicast address and a different port (5353 vs 53).

A fully standards compliant mDNS implementation would use UDP source
port 5353 as well to issue queries, however we MUST NOT use that port
and continue using an ephemeral port until features such as service
discovery are implemented.

This change also allows specifying what kind of DNS query we wish to
perform.

https://www.rfc-editor.org/rfc/rfc6762#section-5.1

benbrittain committed:
Cleans up some clippy lints

benbrittain committed:
Adds DNS query type to example

timokroeger commented on src/socket/dns.rs:
```diff
@@ -234,19 +251,33 @@ impl<'a> Socket<'a> {
+        if mdns {
+            // as per RFC 6762 any DNS query ending in .local. MUST be sent as mdns
+            // so we internally overwrite the servers for any of those queries
+            self.update_servers(&[
```
It looks like this line overrides the dns servers permanently. Which would mean that all following regular DNS queries will fail.

benbrittain replied:
Totally does, fixed.

benbrittain committed:
Don't overwrite the DNS servers list permanently

thvdveld committed:
Fix 6LoWPAN fragmentation

thvdveld commented on src/socket/dns.rs:
```diff
@@ -200,6 +211,12 @@ impl<'a> Socket<'a> {
         let mut raw_name: Vec<u8, MAX_NAME_LEN> = Vec::new();
 
+        let mut mdns = false;
+        if name.split(|&c| c == b'.').last().unwrap() == b"local" {
```
What would happen if the name is just "local", without a dot. Then `.last()` will return "local". Is this correct behaviour? Like I said, I don't know anything about mDNS, so it could be that this is valid.

benbrittain replied:
My understanding is that "local" without anything else refers to the top-level domain, in which case the RFC does say under "3.0 Multicast DNS Names" that this has the special semantics.

Dirbaio commented on src/socket/dns.rs:
```diff
@@ -81,6 +90,7 @@ struct PendingQuery {
     delay: Duration,
 
     server_idx: usize,
+    mdns: bool,
```
This could be a "Protocol" enum, with valus "Dns" and "Mdns".

benbrittain replied:
I don't think the API should encourage thinking about mDNS and DNS queries as fundamentally different protocols

benbrittain replied:
I did something in this vein when I put this behind a feature flag

benbrittain committed:
Move mDNS implementation behind feature flag

Swaps the bool in the raw query API to an enum that can statically
prevent mdns usage without the feature flag enabled.

Dirbaio commented on src/iface/interface.rs:
```diff
@@ -1015,22 +1015,14 @@ impl<'a> Interface<'a> {
         #[cfg(feature = "proto-ipv4-fragmentation")]
-        if let Err(e) = self
-            .fragments
+        self.fragments
```
seems there are some unrelated changes

Dirbaio replied:
ah it's clippy lints, I see

Dirbaio reviewed (approved):
Thank you!

bors r+

(Yes, the enum becoming single-case is fine, it's the clearest way to do this. We do this in other places already, for example for ipv4/v6)
